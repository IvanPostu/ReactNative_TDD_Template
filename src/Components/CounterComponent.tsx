import React, {useEffect, useState} from 'react';
import {Button, StyleSheet, Text, View} from 'react-native';

export const CounterComponent = () => {
  const [count, setCount] = useState(0);

  useEffect(() => {
    setTimeout(() => {
      setCount(101);
    }, 900);
  }, []);

  return (
    <View style={styles.root}>
      <Text testID="out" style={styles.text}>
        {count}
      </Text>
      <Button
        testID="incrBtn"
        onPress={() => setCount(prevState => prevState + 1)}
        title="Increment"
      />
      <Button
        testID="decrBtn"
        onPress={() => setCount(prevState => prevState - 1)}
        title="Decriment"
      />
    </View>
  );
};

const styles = StyleSheet.create({
  root: {
    height: 300,
    backgroundColor: 'grey',
  },
  btn: {
    margin: 20,
  },
  text: {
    fontSize: 70,
  },
});
