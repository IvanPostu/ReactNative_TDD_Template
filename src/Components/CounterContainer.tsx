import {ReduxGlobalStateType} from '@/Redux';
import React, {ReactElement, useState} from 'react';
import {Button, StyleSheet, Text, TextInput, View} from 'react-native';
import {connect} from 'react-redux';
import {bindActionCreators, Dispatch} from 'redux';
import * as actionCreators from '@/Redux/Counter/actionCreators';
import Button1 from './Button';

type CounterContainerPropTypes = {} & ReturnType<typeof mapStateToProps> &
  ReturnType<typeof mapDispatchToProps>;

export const CounterContainer = (
  props: CounterContainerPropTypes,
): ReactElement => {
  const [count, setCount] = useState(0);

  function onCounterTextChange(text: string) {
    try {
      const validNumber = text.replace(/[^0-9]/g, '');
      setCount(+validNumber);
    } catch {}
  }

  return (
    <View style={styles.root}>
      <Text testID="outputValue" style={styles.text}>
        {String(props.count)}
      </Text>
      <View>
        <Button
          testID="incrementBtn"
          onPress={() => props.incrementActionCreator()}
          title="Increment"
        />
        <Button
          testID="decrementBtn"
          onPress={() => props.decrementActionCreator()}
          title="Decriment"
        />
        <TextInput
          testID="inputValue"
          keyboardType="numeric"
          style={styles.input}
          value={String(count)}
          onChangeText={onCounterTextChange}
        />
        <Button
          testID="setValueButton"
          onPress={() => {
            props.setValueActionCreator(count);
          }}
          title="Set value"
        />
        <Button1
          isLoading={false}
          onPress={() => {
            console.log(1);
          }}
          title="qq"
          transparent={false}
        />
      </View>
    </View>
  );
};

const mapStateToProps = (state: ReduxGlobalStateType) => ({
  count: state.counterReducer.counter,
});

function mapDispatchToProps(dispatch: Dispatch) {
  return bindActionCreators(actionCreators, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(CounterContainer);

const styles = StyleSheet.create({
  root: {
    height: 400,
    backgroundColor: 'black',
  },
  text: {
    color: '#fff',
  },
  input: {
    backgroundColor: '#fff',
  },
});
