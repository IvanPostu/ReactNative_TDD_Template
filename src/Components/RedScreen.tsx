import React from 'react';
import {StyleSheet, View} from 'react-native';

export default function RedScreen() {
  return <View style={styles.root} />;
}

const styles = StyleSheet.create({
  root: {
    height: 200,
    backgroundColor: 'red',
  },
});
