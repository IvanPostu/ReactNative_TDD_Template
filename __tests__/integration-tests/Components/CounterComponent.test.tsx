import 'react-native';
import React from 'react';
import {fireEvent, render, act} from '@testing-library/react-native';
import {CounterComponent} from '@/Components/CounterComponent';

describe('CounterComponent test', () => {
  test('CounterComponent snapshot', () => {
    const tree = render(<CounterComponent />);
    expect(tree).toMatchSnapshot();
  });

  test('Increment button click test', () => {
    const rendererComponent = render(<CounterComponent />);
    const incrButton = rendererComponent.getByTestId('incrBtn');

    fireEvent.press(incrButton);
    fireEvent.press(incrButton);
    fireEvent.press(incrButton);
    fireEvent.press(incrButton);

    const outTextElement = rendererComponent.getByTestId('out');

    expect(outTextElement.props.children).toEqual(4);
  });

  test('Timeout is called', () => {
    const rendererComponent = render(<CounterComponent />);
    const outTextElement = rendererComponent.getByTestId('out');

    act(() => {
      jest.runAllTimers();
    });

    expect(outTextElement.props.children).toEqual(101);
  });
});
