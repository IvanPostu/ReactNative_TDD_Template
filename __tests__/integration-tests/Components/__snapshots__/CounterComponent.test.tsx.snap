// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CounterComponent component test Renders correctly (take snapshot) 1`] = `
<View
  style={
    Object {
      "backgroundColor": "grey",
      "height": 300,
    }
  }
>
  <Text
    style={
      Object {
        "fontSize": 70,
      }
    }
    testID="out"
  >
    0
  </Text>
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="incrBtn"
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Increment
      </Text>
    </View>
  </View>
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="decrBtn"
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Decriment
      </Text>
    </View>
  </View>
</View>
`;
